### PR TITLE
Added Rate limiting for App and DB creation

### LINKED ISSUE #226: Add rate limiting to App/DB creation for  users
Limit users to only have a certain number of App/Database instances at a given time, to avoid spam in production

### PR DESCRIPTION
**Pull request by reeshipaul**
Added Rate limiting for App and DB creation
A single user can only create upto 10 App Instances and 10 DB Instances to prevent spamming in production

### CODE REVIEW

**Comment by reeshipaul** [comment]
Added rate limiting variables in config file

**Comment by reeshipaul** [comment]
Rate Limiting is time framed, i.e. an user cannot deploy more than the specified number of App and Db instances in one day

**Comment by reeshipaul** [comment]
Rate Limiting rules do not apply to Admin or SuperUser

File: configs/types.go [hunk 1]
```diff
@@ -49,6 +49,7 @@ type GenericService struct {
 type AppMakerService struct {
 	GenericService
 	MetricsInterval time.Duration `toml:"metrics_interval"`
+	AppLimit 		    int 	  `toml:"app_limit"`
```
Comment (h3llix) [inline configs/types.go]: format this! Hint read contributing guidelines

File: configs/types.go [hunk 2]
```diff
@@ -104,6 +105,7 @@ type DbMakerService struct {
 	MongoDB    DatabaseService `toml:"mongodb"`
 	PostgreSQL DatabaseService `toml:"postgresql"`
 	Redis      DatabaseService `toml:"redis"`
+	DbLimit  			int    `toml:"db_limit"`
```
Comment (h3llix) [inline configs/types.go]: format

File: types/application.go [hunk 3]
```diff
@@ -279,3 +281,8 @@ func (app *ApplicationConfig) SetSSHCmd(port int, appName, IP string) {
 func (app *ApplicationConfig) SetOwner(owner string) {
 	app.Owner = owner
 }
+
+//SetDateTime sets the date on which the db was created
```
Comment (h3llix) [inline types/application.go]: Nit :
```suggestion
// SetDateTime sets the date on which the db was created.
```

File: types/database.go [hunk 4]
```diff
@@ -101,3 +104,8 @@ func (db *DatabaseConfig) SetOwner(owner string) {
 func (db *DatabaseConfig) SetSuccess(success bool) {
 	db.Success = success
 }
+
+//SetDateTime sets the date on which the app was created
```
Comment (h3llix) [inline types/database.go]:
```suggestion
// SetDateTime sets the date on which the app was created
```

File: config.sample.toml [hunk 5]
```diff
@@ -182,6 +182,8 @@ port = 4000
 # Time Interval (in seconds) in which metrics of all application containers
 # running in the current node are collected and stored in the central mongoDB database
 metrics_interval = 600
+app_limit = 10
+# Limits the number of app instances that can be deployed by an user
```
Comment (h3llix) [inline config.sample.toml]: I feel comment should be placed above or next to the variable.

File: config.sample.toml [hunk 6]
```diff
@@ -191,6 +193,8 @@ metrics_interval = 600
 [services.dbmaker]
 deploy = false  # Deploy DbMaker?
 port = 9000
+db_limit= 10
+# Limits the number of db instances that can be deployed by an user
```
Comment (h3llix) [inline config.sample.toml]: Same here.

File: services/dbmaker/controller.go [hunk 7]
```diff
@@ -31,10 +31,24 @@ func (s *server) Create(ctx context.Context, body *pb.RequestBody) (*pb.Response
 		return nil, err
 	}
 
+	filter := types.M{
+		"owner": body.GetOwner(),
+		"instance_type": "database",
+		"date": time.Now().Format("2006-02-01"),
```
Comment (h3llix) [inline services/dbmaker/controller.go]: https://pkg.go.dev/time#pkg-constants. Use the constants here.

Comment (h3llix) [inline services/dbmaker/controller.go]: Maybe you can create a helper function for this and pass instance type to create a filter.

File: services/appmaker/controller.go [hunk 8]
```diff
@@ -32,12 +32,26 @@ func (s *server) Create(ctx context.Context, body *pb.RequestBody) (*pb.Response
 	if err != nil {
 		return nil, err
 	}
+	
+	filter := types.M{
+		"owner": body.GetOwner(),
+		"instance_type": "application",
+		"date": time.Now().Format("2006-02-01"),
```
Comment (h3llix) [inline services/appmaker/controller.go]: Dont hardcode key values. define them as constants. If we ever require to change them then we dont need to change these keys at places in the code.

File: services/appmaker/controller.go [hunk 9]
```diff
@@ -32,12 +32,26 @@ func (s *server) Create(ctx context.Context, body *pb.RequestBody) (*pb.Response
 	if err != nil {
 		return nil, err
 	}
+	
+	filter := types.M{
+		"owner": body.GetOwner(),
+		"instance_type": "application",
```
Comment (h3llix) [inline services/appmaker/controller.go]: Create default constant for application instance type or database instance

File: services/appmaker/controller.go [hunk 10]
```diff
@@ -32,12 +32,26 @@ func (s *server) Create(ctx context.Context, body *pb.RequestBody) (*pb.Response
...
+		"date": time.Now().Format("2006-02-01"),
+	}
+	filter[mongo.AdminKey]="false"
```
Comment (h3llix) [inline services/appmaker/controller.go]: Why not boolean?

**Review by h3llix — CHANGES_REQUESTED** [review]

File: configs/types.go (hunk 2 above)
Comment (reeshipaul) [reply configs/types.go]: done

File: types/database.go (hunk 4 above)
Comment (reeshipaul) [reply types/database.go]: done

File: config.sample.toml (hunk 5 above)
Comment (reeshipaul) [reply config.sample.toml]: done

File: config.sample.toml (hunk 6 above)
Comment (reeshipaul) [reply config.sample.toml]: done

File: services/dbmaker/controller.go (hunk 7 above)
Comment (reeshipaul) [reply services/dbmaker/controller.go]: done

File: services/appmaker/controller.go (hunk 8 above)
Comment (reeshipaul) [reply services/appmaker/controller.go]: done

File: services/appmaker/controller.go (hunk 10 above)
Comment (reeshipaul) [reply services/appmaker/controller.go]: done

File: services/appmaker/controller.go (hunk 9 above)
Comment (reeshipaul) [reply services/appmaker/controller.go]: done

File: services/dbmaker/controller.go (hunk 7 above)
Comment (reeshipaul) [reply services/dbmaker/controller.go]: Added new custom 
constants

File: config.sample.toml [hunk 11]
```diff
@@ -183,6 +183,8 @@ port = 4000
 # running in the current node are collected and stored in the central mongoDB database
 metrics_interval = 600
```
Comment (h3llix) [inline config.sample.toml]: Remove new line.

File: services/appmaker/controller.go [hunk 12]
```diff
@@ -33,11 +33,19 @@ func (s *server) Create(ctx context.Context, body *pb.RequestBody) (*pb.Response
 		return nil, err
 	}
 
+	appCount := c.CountInstancesOfCurrentDate(body.GetOwner(), mongo.AppInstance)
+	fmt.Printf("App count %v", appCount)
```
Comment (h3llix) [inline services/appmaker/controller.go]: Do we really need to print this? Or was it for testing purposes.

**Review by h3llix — CHANGES_REQUESTED** [review]

File: services/master/controllers/helper.go [hunk 13]
```diff
@@ -48,6 +48,16 @@ func fetchInstances(c *gin.Context, instance string) {
...
+// CountInstancesOfCurrentDate counts the instances of the given type owned by the user
+// func CountInstancesOfCurrentDate(owner string, instanceType string) int {
+// 	filter := make(types.M)
+// 	filter[mongo.InstanceTypeKey] = instanceType
+// 	filter[mongo.OwnerKey] = owner
+// 	filter[mongo.AdminKey] = "false"
+// 	apps := mongo.FetchInstances(filter)
+// 	return len(apps)
+// }
+
```
Comment (h3llix) [inline services/master/controllers/helper.go]: Remove these

File: config.sample.toml [hunk 14]
```diff
@@ -182,7 +182,12 @@ port = 4000
 # Time Interval (in seconds) in which metrics of all application containers
 # running in the current node are collected and stored in the central mongoDB database
 metrics_interval = 600
-
+# Limits the number of app instances that can be deployed by an user
```
Comment (h3llix) [inline config.sample.toml]:
```suggestion
# Limits the total number of app instances that can be deployed by an user
```

File: config.sample.toml [hunk 15]
```diff
@@ -182,7 +182,12 @@ port = 4000
...
+# Set app_limit = -1 if no limit is to be imposed
+app_limit = 10
+# Time interval (in hours) for rate limiting
```
Comment (h3llix) [inline config.sample.toml]: The comment is not really informative. We should also inform that x instances can be deployed in this time interval but the total number of instances should not be more than `app_limit`

File: config.sample.toml [hunk 16]
```diff
@@ -191,6 +196,9 @@ metrics_interval = 600
 [services.dbmaker]
 deploy = false  # Deploy DbMaker?
 port = 9000
+# Limits the number of db instances that can be deployed by an user
+# Set db_limit = -1 if no limit is to be imposed
+db_limit= 10
```
Comment (h3llix) [inline config.sample.toml:206]: No rate limiting for DB instances ?

File: configs/types.go [hunk 17]
```diff
@@ -104,6 +106,7 @@ type DbMakerService struct {
 	MongoDB    DatabaseService `toml:"mongodb"`
 	PostgreSQL DatabaseService `toml:"postgresql"`
 	Redis      DatabaseService `toml:"redis"`
+	DbLimit    int             `toml:"db_limit"`
```
Comment (h3llix) [inline configs/types.go]:
```suggestion
	DBLimit    int             `toml:"db_limit"`
```

Comment (h3llix) [reply configs/types.go]: Just a suggestion.

**Review by h3llix — APPROVED** [review]

**Comment by h3llix** [comment]
LGTM!

**Comment by aryaman-titan** [comment]
LGTM as well! Tested the PR

**Pull request merged**